pkg/kademlia/workers.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package kademlia
+
+import (
+	"bytes"
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/node"
+	proto "storj.io/storj/protos/overlay"
+)
+
+const (
+	uncontacted = iota
+	inProgress
+	completed
+)
+
+type chore struct {
+	status int
+	node   *proto.Node
+}
+
+type worker struct {
+	// node id to chore
+	contacted   map[string]*chore
+	mu          *sync.Mutex
+	maxResponse time.Duration
+	cancel      context.CancelFunc
+	nodeClient  node.Client
+	find        proto.Node
+	k           int
+}
+
+func (w *worker) work(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			w.mu.Lock()
+			var wk *chore
+			for i, v := range w.contacted {
+				if v.status == uncontacted {
+					wk = v
+					w.contacted[i].status = inProgress
+					w.mu.Unlock()
+					break
+				}
+			}
+
+			if wk == nil {
+				w.mu.Unlock()
+				time.AfterFunc(2*w.maxResponse, w.cancel)
+				return nil
+			}
+
+			start := time.Now()
+			nodes, err := w.nodeClient.Lookup(ctx, *wk.node, w.find)
+			if err != nil {
+				return err
+			}
+			latency := time.Now().Sub(start)
+			if latency > w.maxResponse {
+				w.maxResponse = latency
+			}
+
+			w.mu.Lock()
+			if (len(w.contacted) + len(nodes)) < w.k {
+				for _, v := range nodes {
+					w.contacted[v.GetId()] = &chore{node: v, status: uncontacted}
+				}
+				w.mu.Unlock()
+			}
+
+			// sort all nodes from map
+			mapNodes := make([][]byte, len(w.contacted))
```

Comment:
so for all of this logic, perhaps you could simplify it if you used container/heap and made a heap-based priority queue, where you can add nodes and then the heap lets you pop nodes that are farthest away. that way you just add and pop nodes when you hear about new nodes, and the closest nodes stay in the data structure?